pkg/mutation/mutation_function.go

```diff
@@ -0,0 +1,202 @@
+package mutation
+
+import (
+	"fmt"
+
+	"github.com/open-policy-agent/gatekeeper/pkg/mutation/path/parser"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+func Mutate(mutator Mutator, obj *unstructured.Unstructured) error {
+	return mutate(mutator, obj.Object, nil, 0)
+}
+
+func mutate(m Mutator, current interface{}, previous interface{}, depth int) error {
+	if len(m.Path().Nodes)-1 == depth {
+		return addValue(m, current, previous, depth)
+	}
+	pathEntry := m.Path().Nodes[depth]
+	switch t := pathEntry.Type(); t {
+	case parser.ObjectNode:
+		objectPathEntry := pathEntry.(*parser.Object)
```

Comment:
should we check for error returned by type casting here?

Reply:
The pathEntry.Type() function (used in the switch) points to the type of Node we have. 
`case parser.ObjectNode:` means the pathEntry is of type `parser.Object`.
Imho skipping the check here is safe, as this error would require a hard change in the code (and releasing without running tests). It is impossible to get this cast to fail by changing some runtime parameters.

Reply:
+1

it's slightly safer if we did:

```
t := pathEntry.Type()
switch pathEntry.(type) {
case *parser.Object:
...
...
...
}
```

but there we're brittle to parser returning objects instead of pointers.

If we have tests that cover both object nodes and list nodes (which I think we do), we'd catch any possible panics. Safe type-casting was definitely necessary when we were dealing with the input object, as there we have no control over what the user is sending.

Reply:
To properly test this (get a casting error), we would have to change the implementation of the parser.Object and parser.List.
Note that the Type() method has a hardcoded return value. For any other type we return an error

Reply:
Ah, not suggesting we need to specifically test for casting errors. I was just saying that if we accidentally change the parser code such that we would see casting errors, the panic would show up in the unit tests we currently have.

Reply:
Agree with this. Current test would definitely not work if we change this code.

Reply:
Changes all the switch statements to use type instead of the Type() function.

Reply:
@ritazh does this change LGTY now?